units/host_termination.go

```diff
@@ -0,0 +1,190 @@
+package units
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/evergreen-ci/evergreen"
+	"github.com/evergreen-ci/evergreen/cloud"
+	"github.com/evergreen-ci/evergreen/model/event"
+	"github.com/evergreen-ci/evergreen/model/host"
+	"github.com/evergreen-ci/evergreen/notify"
+	"github.com/evergreen-ci/evergreen/util"
+	"github.com/mongodb/amboy"
+	"github.com/mongodb/amboy/dependency"
+	"github.com/mongodb/amboy/job"
+	"github.com/mongodb/grip"
+	"github.com/mongodb/grip/message"
+	"github.com/pkg/errors"
+)
+
+const hostTerminationJobName = "host-termination-job"
+
+type hostTerminationJob struct {
+	HostID   string `bson:"host_id" json:"host_id" yaml:"host_id"`
+	job.Base `bson:"metadata" json:"metadata" yaml:"metadata"`
+
+	host *host.Host
+	env  evergreen.Environment
+}
+
+func makeHostTerminationJob() *hostTerminationJob {
+	j := &hostTerminationJob{
+		Base: job.Base{
+			JobType: amboy.JobType{
+				Name:    hostTerminationJobName,
+				Version: 0,
+			},
+		},
+	}
+
+	j.SetDependency(dependency.NewAlways())
+	return j
+}
+
+func NewHostTerminationJob(env evergreen.Environment, h host.Host) amboy.Job {
+	j := makeHostTerminationJob()
+	j.host = &h
+	j.HostID = h.Id
+	j.env = env
+	j.SetPriority(2)
+	j.SetID(fmt.Sprintf("%s.%s", hostTerminationJobName, j.HostID))
+
+	return j
+}
+
+func (j *hostTerminationJob) Run(ctx context.Context) {
+	var err error
+
+	if j.host == nil {
+		j.host, err = host.FindOneId(j.HostID)
+		if err != nil {
+			j.AddError(err)
+			return
+		}
+	}
+
+	if j.env == nil {
+		j.env = evergreen.GetEnvironment()
+	}
+
+	settings := j.env.Settings()
+
+	if !util.StringSliceContains(evergreen.UphostStatus, j.host.Status) {
+		// if host isn't in a running state we shouldn't try to terminate it.
+		grip.Debug(message.Fields{
+			"job":     j.ID(),
+			"host":    j.HostID,
+			"status":  j.host.Status,
+			"message": "host not running, termination job is a noop",
+		})
+		return
+	}
+
+	idleTimeStartsAt := j.host.LastTaskCompletedTime
+	if idleTimeStartsAt.IsZero() || idleTimeStartsAt == util.ZeroTime {
+		idleTimeStartsAt = j.host.StartTime
+	}
+
+	// clear the running task of the host in case one has been assigned.
+	if j.host.RunningTask != "" {
+		grip.Warning(message.Fields{
+			"runner":   "monitor",
```

Comment:
Same here: should we log job type rather than monitor?

Reply:
done